Libs/Mesh/Mesh.cpp

```diff
@@ -537,16 +535,52 @@ int Mesh::closestPointId(const Point3 point)
 
 double Mesh::geodesicDistance(int source, int target)
 {
-  Eigen::MatrixXd V = points();
-  Eigen::MatrixXi F = faces();
-  Eigen::VectorXd VS(1), VT(1);
-  VS[0] = source;
-  VT[0] = target;
-  Eigen::VectorXd FS, FT;
-  Eigen::VectorXd d;
-
-  igl::exact_geodesic(V,F,VS,FS,VT,FT,d);
-  return d[0];
+  VtkMeshWrapper wrap(this->mesh, true);
+  double distance = wrap.ComputeDistance(source, -1, target, -1);
+
+  return distance;
+}
+
+Field Mesh::geodesicDistance(const Point3 landmark)
+{
+  vtkSmartPointer<vtkDoubleArray> distance = vtkSmartPointer<vtkDoubleArray>::New();
+  distance->SetNumberOfComponents(1);
+  distance->SetNumberOfTuples(numPoints());
+  distance->SetName("GeodesicDistanceToLandmark");
+
+  VtkMeshWrapper wrap(this->mesh, true);
+
+  for (int i = 0; i < numPoints(); i++)
+  {
+    distance->SetValue(i, wrap.ComputeDistance(getPoint(i), -1, landmark, -1));
```

Comment:
tl;dr: `wrap.ComputeDistance(landmark, 0, getPoint(i), -1)`

This doesn't take advantage the MeshWrapper caching for the first point (or any others). I suggest passing landmark w/ idx=0 for the first point. From my perusal of the vtkMeshWrapper caching system, this should only allocate one item in the underlying cache, used by ComputeFaceAndWeights, and which still exists even though it's noted to have no performance advantage. So we want to only send idx 0 to minimize the "unused-but-still-used" cache, but still enable the first point cache, which is handled in a different way (in both ComputeDistance and IsWithinDistance).

Reply:
What's the "unused-but-still-used" cache?

Reply:
The function `GetTriangleForPoint` performs caching based on the provided index. It is called by `ComputeFaceAndWeights`. Since we are deliberately only passing -1 as the index for points other than the landmark (following my recommendation here, because first point has a different cache mechanism, explicitly associating the face id and barycentric weights with the first index passed), the mechanism in the `GetTriangleForPoint` function (an array of indices associated with the triangle in which they belong) will never be used. However, because it still allocates an array of points to associate with an index, using index 0 for the first point is the best way to limit this unnecessary allocation (short of reorganizing the mesh wrapper itself).

Reply:
If you always pass idx=-1, the `particle_triangles_` cache won't ever be allocated.

Reply:
> Correct. But you still want to cache the first point, which is why you want to swap the order of arguments and pass 0 as the first argument’s (the source pt) cache index. The dest point can keep using -1, and should, since according to the comments in the code, only caching the first point made a difference.

I agree that using '0' and landmark as the first point in this case is very slightly better, but I want to make sure the caches and the comments in the code are understood.  ComputeDistance has a separate **single** cached value of only the very last query such that `ComputeFaceAndWeights` and `GetTriangleForPoint` are not even called (nor is the `particle_triangles_` cache even used).  For optimal speed, Karthik was trying to avoid even the `IsInTriangle` check that happens when the `particle_triangles_` cache is used.  He didn't find that a secondary single value cache of the 2nd point was helpful because the optimizer never calls with the 2nd same point twice in a row.  This is true of the `Field Mesh::geodesicDistance(const Point3 landmark)` function as well.  But these comments are not referring to the cache in `GetTriangleForPoint`.  That cache is still heavily used by the optimizer for the 2nd point, but it will call `IsInTriangle` to confirm validity.

Keep in mind that if someone combines the Mesh and VtkMeshWrapper code, always using idx=0 and -1 will lead to incorrect results when a user's code makes repeated `Mesh::geodesicDistance` with different landmarks and these are always passed in as `idx=0` and the last query check matches even though the point's position is different.

This won't be a problem in this code now because it creates a new VtkMeshWrapper object for each call.  As a side note, have a look at the constructor of VtkMeshWrapper and see it is probably millions of times slower than ComputeDistance calls.  If you are concerned about performance here, you should keep around the VtkMeshWrapper.

Reply:
> Keep in mind that if someone combines the Mesh and VtkMeshWrapper code, always using idx=0 and -1 will lead to incorrect results when a user's code makes repeated Mesh::geodesicDistance with different landmarks and these are always passed in as idx=0 and the last query check matches even though the point's position is different.

Thank you. That's a very useful issue to identify since as written it will only work for the first computation of geodesic distance from a point and any other points will use invalid cache values. It should be changed to check the values of the point instead of its index. This addresses both issue at once: no `particle_triange_` cache will be allocated, and the omission of calling `GetTriangleForPoint` will remain. Reuse of idx will no longer accidentally result in an incorrect face id and barycentric coordinates being returned for the first cached point*.

*It's technically possible, but the same point will always return the same triangle and barycentric coordinates in that triangle for the same wrapped mesh, and there is no way to change which mesh is wrapped.

Reply:
The cut and pasted use of the "first point cache" is also troublesome, but can easily be addressed. 
The `particle_trangles_` cache is never used without providing idx, and (with the safety updates above) it will never be provided.

Reply:
The latest push dumps caching of the first point altogether. Turns out swapping the argument order speeds up computation by _more than 600%_, but it's due to internal caching of the vtkCellLocator (perhaps the same trick).